src/relay/backend/vm/compiler.cc

```diff
@@ -605,21 +650,49 @@ class VMFunctionCompiler : ExprFunctor<void(const Expr& expr)> {
       // heterogeneous execution.
       LOG(FATAL) << "Currently VM compiler doesn't support heterogeneous compilation";
     }
-    auto key = CCacheKeyNode::make(func, target);
-    auto cfunc = engine_->Lower(key);
-    // TODO(jroesch): support lowered funcs for multiple targets
-    CHECK_EQ(cfunc->funcs.size(), 1);
-    auto op_index = -1;
-    if (context_->seen_funcs.find(cfunc->funcs[0]) == context_->seen_funcs.end()) {
-      op_index = context_->cached_funcs.size();
-      context_->cached_funcs.push_back(cfunc);
-      context_->seen_funcs[cfunc->funcs[0]] = op_index;
+
+    if (!need_jit) {
+      auto key = CCacheKeyNode::make(func, target);
+      auto cfunc = engine_->Lower(key);
+      // TODO(jroesch): support lowered funcs for multiple targets
+      CHECK_EQ(cfunc->funcs.size(), 1);
+      auto cfunc_index = GetCFuncIndex(cfunc);
+      Emit(Instruction::InvokePacked(cfunc_index, arity, return_count, unpacked_arg_regs));
     } else {
-      op_index = context_->seen_funcs[cfunc->funcs[0]];
+      using JITMap = std::map<InlineCacheKey, PackedFunc>;
+      auto jit_map = JITMap();
+      auto type = Downcast<FuncType>(func->checked_type());
+      auto pf = PackedFunc([=](TVMArgs args, TVMRetValue* rv) mutable {
+        InlineCacheKey ick;
+        for (size_t i = 0; i < arity; ++i) {
+          NDArray arr = args[i];
+          UpdateICK(&ick, arr);
+        }
+        tvm::Map<Var, Expr> bindings;
+        if (jit_map.count(ick) == 0) {
+          auto func_type = Downcast<FuncType>(func->checked_type());
+          for (size_t i = 0; i < func->params.size(); ++i) {
+            auto int_shape = NDArray(args[i]).Shape();
+            tvm::Array<tvm::Expr> shape;
+            for (auto s : int_shape) {
+              shape.push_back(tvm::Expr(static_cast<int32_t>(s)));
+            }
+            auto dtype = Downcast<TensorType>(func_type->arg_types[i])->dtype;
+            bindings.Set(func->params[i],
+                         VarNode::make("new_var", TensorTypeNode::make(shape, dtype)));
+          }
+          auto new_func_untyped = Downcast<Function>(Subst(func, bindings));
+          auto new_func = Downcast<Function>(InferType(new_func_untyped, Module(), GlobalVar()));
+          auto key = CCacheKeyNode::make(new_func, target);
+          CompileEngine ce = CompileEngine::Global();  // WHY cant I use engine_?
+          auto jit_pf = ce->JIT(key);
```

Comment:
Just a thought - would it be worthwhile including an option for users to specify a range for each dimension of a shape that can be `Any`, then we can pre-compile all of those? It will use more memory, but save latency at runtime. 

Alternatively, we can have the user supply a max value for each dimension that can be `Any`, then use that max value everywhere and pad when necessary. Thoughts?

Reply:
I'm inclined to leave it as future work - it is indeed possible, however they are more complex, and have different options (range? max value and pad?). Meanwhile, this make it possible to compute for dense, is simple and allow further extension that does those. Sounds good to you?

Reply:
Definitely, I think we can leave it for a later change.